core/src/main/java/org/apache/calcite/rel/rules/materialize/MaterializedViewAggregateRule.java

```diff
@@ -292,10 +298,36 @@
                 otherCompensationPred)));
 
     // Generate query rewriting.
-    RelNode rewrittenPlan = relBuilder
-        .push(target)
-        .filter(simplify.simplifyUnknownAsFalse(queryCompensationPred))
-        .build();
+    RelNode rewrittenPlan = null;
+    if (pushQueryFilter) {
+      rewrittenPlan = target.accept(
+          // Push the queryCompensationPred down into any filter that is present in the plan
+          new RelShuttleImpl() {
+            @Override public RelNode visit(LogicalFilter filter) {
+              RexNode condition =
+                  RexUtil.flatten(rexBuilder,
+                      rexBuilder.makeCall(
+                          SqlStdOperatorTable.AND,
+                          filter.getCondition(),
+                          queryCompensationPred));
+              return filter.copy(filter.getTraitSet(), filter.getInput(), condition);
+            }
+
+            @Override public RelNode visit(RelNode other) {
+              if (other instanceof RelSubset) {
+                RelSubset relSubset = (RelSubset) other;
+                return relSubset.getBestOrOriginal().accept(this);
+              } else {
+                return visitChildren(other);
+              }
+            }
+          });
+    } else {
+      rewrittenPlan = relBuilder
+          .push(target)
```

Comment:
I think it's worth extracting the `RelShuttleImpl` to a private class to improve readability

Reply:
done